// pjsSystem/windowsMessages.js


exports.WM_NULL             = 0x0000;
exports.WM_CREATE           = 0x0001;
exports.WM_DESTROY          = 0x0002;
exports.WM_MOVE             = 0x0003;
exports.WM_SIZE             = 0x0005;
exports.WM_ACTIVATE         = 0x0006;
exports.WM_SETFOCUS         = 0x0007;
exports.WM_KILLFOCUS        = 0x0008;
exports.WM_ENABLE           = 0x000A;
exports.WM_SETREDRAW        = 0x000B;
exports.WM_SETTEXT          = 0x000C;
exports.WM_GETTEXT          = 0x000D;
exports.WM_GETTEXTLENGTH    = 0x000E;
exports.WM_PAINT            = 0x000F;
exports.WM_CLOSE            = 0x0010;
exports.WM_QUERYENDSESSION  = 0x0011;
exports.WM_QUIT             = 0x0012;
exports.WM_QUERYOPEN        = 0x0013;
exports.WM_ERASEBKGND       = 0x0014;
exports.WM_SYSCOLORCHANGE   = 0x0015;
exports.WM_ENDSESSION       = 0x0016;
exports.WM_SYSTEMERROR      = 0x0017;
exports.WM_SHOWWINDOW       = 0x0018;
exports.WM_CTLCOLOR         = 0x0019;
exports.WM_WININICHANGE     = 0x001A;
exports.WM_SETTINGCHANGE = exports.WM_WININICHANGE;
exports.WM_DEVMODECHANGE    = 0x001B;
exports.WM_ACTIVATEAPP      = 0x001C;
exports.WM_FONTCHANGE       = 0x001D;
exports.WM_TIMECHANGE       = 0x001E;
exports.WM_CANCELMODE       = 0x001F;
exports.WM_SETCURSOR        = 0x0020;
exports.WM_MOUSEACTIVATE    = 0x0021;
exports.WM_CHILDACTIVATE    = 0x0022;
exports.WM_QUEUESYNC        = 0x0023;
exports.WM_GETMINMAXINFO    = 0x0024;
exports.WM_PAINTICON        = 0x0026;
exports.WM_ICONERASEBKGND   = 0x0027;
exports.WM_NEXTDLGCTL       = 0x0028;
exports.WM_SPOOLERSTATUS    = 0x002A;
exports.WM_DRAWITEM         = 0x002B;
exports.WM_MEASUREITEM      = 0x002C;
exports.WM_DELETEITEM       = 0x002D;
exports.WM_VKEYTOITEM       = 0x002E;
exports.WM_CHARTOITEM       = 0x002F;
exports.WM_SETFONT          = 0x0030;
exports.WM_GETFONT          = 0x0031;
exports.WM_SETHOTKEY        = 0x0032;
exports.WM_GETHOTKEY        = 0x0033;
exports.WM_QUERYDRAGICON    = 0x0037;
exports.WM_COMPAREITEM      = 0x0039;
exports.WM_GETOBJECT        = 0x003D;
exports.WM_COMPACTING       = 0x0041;
exports.WM_COMMNOTIFY       = 0x0044;    
exports.WM_WINDOWPOSCHANGING = 0x0046;
exports.WM_WINDOWPOSCHANGED = 0x0047;
exports.WM_POWER            = 0x0048;
exports.WM_COPYDATA         = 0x004A;
exports.WM_CANCELJOURNAL    = 0x004B;
exports.WM_NOTIFY           = 0x004E;
exports.WM_INPUTLANGCHANGEREQUEST = 0x0050;
exports.WM_INPUTLANGCHANGE  = 0x0051;
exports.WM_TCARD            = 0x0052;
exports.WM_HELP             = 0x0053;
exports.WM_USERCHANGED      = 0x0054;
exports.WM_NOTIFYFORMAT     = 0x0055;
exports.WM_CONTEXTMENU      = 0x007B;
exports.WM_STYLECHANGING    = 0x007C;
exports.WM_STYLECHANGED     = 0x007D;
exports.WM_DISPLAYCHANGE    = 0x007E;
exports.WM_GETICON          = 0x007F;
exports.WM_SETICON          = 0x0080;
exports.WM_NCCREATE         = 0x0081;
exports.WM_NCDESTROY        = 0x0082;
exports.WM_NCCALCSIZE       = 0x0083;
exports.WM_NCHITTEST        = 0x0084;
exports.WM_NCPAINT          = 0x0085;
exports.WM_NCACTIVATE       = 0x0086;
exports.WM_GETDLGCODE       = 0x0087;
exports.WM_NCMOUSEMOVE      = 0x00A0;
exports.WM_NCLBUTTONDOWN    = 0x00A1;
exports.WM_NCLBUTTONUP      = 0x00A2;
exports.WM_NCLBUTTONDBLCLK  = 0x00A3;
exports.WM_NCRBUTTONDOWN    = 0x00A4;
exports.WM_NCRBUTTONUP      = 0x00A5;
exports.WM_NCRBUTTONDBLCLK  = 0x00A6;
exports.WM_NCMBUTTONDOWN    = 0x00A7;
exports.WM_NCMBUTTONUP      = 0x00A8;
exports.WM_NCMBUTTONDBLCLK  = 0x00A9;
exports.WM_NCXBUTTONDOWN = 0x00AB;
exports.WM_NCXBUTTONUP = 0x00AC;
exports.WM_NCXBUTTONDBLCLK = 0x00AD;
exports.WM_INPUT_DEVICE_CHANGE = 0x00FE;
exports.WM_INPUT = 0x00FF;
exports.WM_KEYFIRST         = 0x0100;
exports.WM_KEYDOWN          = 0x0100;
exports.WM_KEYUP            = 0x0101;
exports.WM_CHAR             = 0x0102;
exports.WM_DEADCHAR         = 0x0103;
exports.WM_SYSKEYDOWN       = 0x0104;
exports.WM_SYSKEYUP         = 0x0105;
exports.WM_SYSCHAR          = 0x0106;
exports.WM_SYSDEADCHAR      = 0x0107;
exports.WM_UNICHAR          = 0x0109;
exports.WM_KEYLAST          = 0x0109;
exports.WM_INITDIALOG       = 0x0110;
exports.WM_COMMAND          = 0x0111;
exports.WM_SYSCOMMAND       = 0x0112;
exports.WM_TIMER            = 0x0113;
exports.WM_HSCROLL          = 0x0114;
exports.WM_VSCROLL          = 0x0115;
exports.WM_INITMENU         = 0x0116;
exports.WM_INITMENUPOPUP    = 0x0117;
exports.WM_GESTURE = 0x0119;
exports.WM_GESTURENOTIFY = 0x011A;
exports.WM_MENUSELECT       = 0x011F;
exports.WM_MENUCHAR         = 0x0120;
exports.WM_ENTERIDLE        = 0x0121;
exports.WM_MENURBUTTONUP    = 0x0122;
exports.WM_MENUDRAG         = 0x0123;
exports.WM_MENUGETOBJECT    = 0x0124;
exports.WM_UNINITMENUPOPUP  = 0x0125;
exports.WM_MENUCOMMAND      = 0x0126;
exports.WM_CHANGEUISTATE    = 0x0127;
exports.WM_UPDATEUISTATE    = 0x0128;
exports.WM_QUERYUISTATE     = 0x0129;
exports.WM_CTLCOLORMSGBOX   = 0x0132;
exports.WM_CTLCOLOREDIT     = 0x0133;
exports.WM_CTLCOLORLISTBOX  = 0x0134;
exports.WM_CTLCOLORBTN      = 0x0135;
exports.WM_CTLCOLORDLG      = 0x0136;
exports.WM_CTLCOLORSCROLLBAR= 0x0137;
exports.WM_CTLCOLORSTATIC   = 0x0138;
exports.WM_MOUSEFIRST       = 0x0200;
exports.WM_MOUSEMOVE        = 0x0200;
exports.WM_LBUTTONDOWN      = 0x0201;
exports.WM_LBUTTONUP        = 0x0202;
exports.WM_LBUTTONDBLCLK    = 0x0203;
exports.WM_RBUTTONDOWN      = 0x0204;
exports.WM_RBUTTONUP        = 0x0205;
exports.WM_RBUTTONDBLCLK    = 0x0206;
exports.WM_MBUTTONDOWN      = 0x0207;
exports.WM_MBUTTONUP        = 0x0208;
exports.WM_MBUTTONDBLCLK    = 0x0209;
exports.WM_MOUSEWHEEL       = 0x020A;
exports.WM_XBUTTONDOWN = 0x020B; 
exports.WM_XBUTTONUP = 0x020C; 
exports.WM_XBUTTONDBLCLK = 0x020D; 
exports.WM_MOUSEHWHEEL = 0x020E; 
exports.WM_MOUSELAST = 0x020E; 
exports.WM_PARENTNOTIFY     = 0x0210;
exports.WM_ENTERMENULOOP    = 0x0211;
exports.WM_EXITMENULOOP     = 0x0212;
exports.WM_NEXTMENU         = 0x0213;
exports.WM_SIZING           = 532;
exports.WM_CAPTURECHANGED   = 533;
exports.WM_MOVING           = 534;
exports.WM_POWERBROADCAST   = 536;
exports.WM_DEVICECHANGE     = 537;
exports.WM_IME_STARTCOMPOSITION        = 0x010D;
exports.WM_IME_ENDCOMPOSITION          = 0x010E;
exports.WM_IME_COMPOSITION             = 0x010F;
exports.WM_IME_KEYLAST                 = 0x010F;
exports.WM_IME_SETCONTEXT              = 0x0281;
exports.WM_IME_NOTIFY                  = 0x0282;
exports.WM_IME_CONTROL                 = 0x0283;
exports.WM_IME_COMPOSITIONFULL         = 0x0284;
exports.WM_IME_SELECT                  = 0x0285;
exports.WM_IME_CHAR                    = 0x0286;
exports.WM_IME_REQUEST                 = 0x0288;
exports.WM_IME_KEYDOWN                 = 0x0290;
exports.WM_IME_KEYUP                   = 0x0291;
exports.WM_MDICREATE        = 0x0220;
exports.WM_MDIDESTROY       = 0x0221;
exports.WM_MDIACTIVATE      = 0x0222;
exports.WM_MDIRESTORE       = 0x0223;
exports.WM_MDINEXT          = 0x0224;
exports.WM_MDIMAXIMIZE      = 0x0225;
exports.WM_MDITILE          = 0x0226;
exports.WM_MDICASCADE       = 0x0227;
exports.WM_MDIICONARRANGE   = 0x0228;
exports.WM_MDIGETACTIVE     = 0x0229;
exports.WM_MDISETMENU       = 0x0230;
exports.WM_ENTERSIZEMOVE    = 0x0231;
exports.WM_EXITSIZEMOVE     = 0x0232;
exports.WM_DROPFILES        = 0x0233;
exports.WM_MDIREFRESHMENU   = 0x0234;
exports.WM_TOUCH            = 0x0240;
exports.WM_MOUSEHOVER       = 0x02A1;
exports.WM_MOUSELEAVE       = 0x02A3;
exports.WM_NCMOUSEHOVER     = 0x02A0;
exports.WM_NCMOUSELEAVE     = 0x02A2;
exports.WM_WTSSESSION_CHANGE = 0x02B1;
exports.WM_TABLET_FIRST     = 0x02C0;
exports.WM_TABLET_LAST      = 0x02DF;
exports.WM_CUT              = 0x0300;
exports.WM_COPY             = 0x0301;
exports.WM_PASTE            = 0x0302;
exports.WM_CLEAR            = 0x0303;
exports.WM_UNDO             = 0x0304;
exports.WM_RENDERFORMAT     = 0x0305;
exports.WM_RENDERALLFORMATS = 0x0306;
exports.WM_DESTROYCLIPBOARD = 0x0307;
exports.WM_DRAWCLIPBOARD    = 0x0308;
exports.WM_PAINTCLIPBOARD   = 0x0309;
exports.WM_VSCROLLCLIPBOARD = 0x030A;
exports.WM_SIZECLIPBOARD    = 0x030B;
exports.WM_ASKCBFORMATNAME  = 0x030C;
exports.WM_CHANGECBCHAIN    = 0x030D;
exports.WM_HSCROLLCLIPBOARD = 0x030E;
exports.WM_QUERYNEWPALETTE  = 0x030F;
exports.WM_PALETTEISCHANGING= 0x0310;
exports.WM_PALETTECHANGED   = 0x0311;
exports.WM_HOTKEY           = 0x0312;
exports.WM_PRINT            = 791;
exports.WM_PRINTCLIENT      = 792;
exports.WM_APPCOMMAND       = 0x0319;
exports.WM_THEMECHANGED     = 0x031A;
exports.WM_CLIPBOARDUPDATE  = 0x031D;
exports.WM_HANDHELDFIRST    = 856;
exports.WM_HANDHELDLAST     = 863;
exports.WM_PENWINFIRST      = 0x0380;
exports.WM_PENWINLAST       = 0x038F;
exports.WM_COALESCE_FIRST   = 0x0390;
exports.WM_COALESCE_LAST    = 0x039F;
exports.WM_DDE_FIRST        = 0x03E0;
exports.WM_DDE_INITIATE     = exports.WM_DDE_FIRST + 0;
exports.WM_DDE_TERMINATE    = exports.WM_DDE_FIRST + 1;
exports.WM_DDE_ADVISE       = exports.WM_DDE_FIRST + 2;
exports.WM_DDE_UNADVISE     = exports.WM_DDE_FIRST + 3;
exports.WM_DDE_ACK          = exports.WM_DDE_FIRST + 4;
exports.WM_DDE_DATA         = exports.WM_DDE_FIRST + 5;
exports.WM_DDE_REQUEST      = exports.WM_DDE_FIRST + 6;
exports.WM_DDE_POKE         = exports.WM_DDE_FIRST + 7;
exports.WM_DDE_EXECUTE      = exports.WM_DDE_FIRST + 8;
exports.WM_DDE_LAST         = exports.WM_DDE_FIRST + 8;
exports.WM_DWMCOMPOSITIONCHANGED        = 0x031E; 
exports.WM_DWMNCRENDERINGCHANGED        = 0x031F;
exports.WM_DWMCOLORIZATIONCOLORCHANGED  = 0x0320;
exports.WM_DWMWINDOWMAXIMIZEDCHANGE     = 0x0321;
exports.WM_DWMSENDICONICTHUMBNAIL       = 0x0323;
exports.WM_DWMSENDICONICLIVEPREVIEWBITMAP = 0x0326;
exports.WM_GETTITLEBARINFOEX = 0x033F;
exports.WM_TABLET_DEFBASE = 0x02C0;
exports.WM_TABLET_MAXOFFSET = 0x20;
exports.WM_TABLET_ADDED = exports.WM_TABLET_DEFBASE + 8;
exports.WM_TABLET_DELETED = exports.WM_TABLET_DEFBASE + 9;
exports.WM_TABLET_FLICK = exports.WM_TABLET_DEFBASE + 11;
exports.WM_TABLET_QUERYSYSTEMGESTURESTATUS = exports.WM_TABLET_DEFBASE + 12;
exports.WM_APP = 0x8000;
exports.WM_USER             = 0x0400;

exports.BN_CLICKED       = 0;
exports.BN_PAINT         = 1;
exports.BN_HILITE        = 2;
exports.BN_UNHILITE      = 3;
exports.BN_DISABLE       = 4;
exports.BN_DOUBLECLICKED = 5;
exports.BN_PUSHED = exports.BN_HILITE;
exports.BN_UNPUSHED = exports.BN_UNHILITE;
exports.BN_DBLCLK = exports.BN_DOUBLECLICKED;
exports.BN_SETFOCUS = 6;
exports.BN_KILLFOCUS = 7;

exports.BM_GETCHECK = 0x00F0;
exports.BM_SETCHECK = 0x00F1;
exports.BM_GETSTATE = 0x00F2;
exports.BM_SETSTATE = 0x00F3;
exports.BM_SETSTYLE = 0x00F4;
exports.BM_CLICK    = 0x00F5;
exports.BM_GETIMAGE = 0x00F6;
exports.BM_SETIMAGE = 0x00F7;
exports.BM_SETDONTCLICK = 0x00F8;

exports.LBN_ERRSPACE  = (-2);
exports.LBN_SELCHANGE = 1;
exports.LBN_DBLCLK    = 2;
exports.LBN_SELCANCEL = 3;
exports.LBN_SETFOCUS  = 4;
exports.LBN_KILLFOCUS = 5;

exports.LB_ADDSTRING            = 0x0180;
exports.LB_INSERTSTRING         = 0x0181;
exports.LB_DELETESTRING         = 0x0182;
exports.LB_SELITEMRANGEEX       = 0x0183;
exports.LB_RESETCONTENT         = 0x0184;
exports.LB_SETSEL               = 0x0185;
exports.LB_SETCURSEL            = 0x0186;
exports.LB_GETSEL               = 0x0187;
exports.LB_GETCURSEL            = 0x0188;
exports.LB_GETTEXT              = 0x0189;
exports.LB_GETTEXTLEN           = 0x018A;
exports.LB_GETCOUNT             = 0x018B;
exports.LB_SELECTSTRING         = 0x018C;
exports.LB_DIR                  = 0x018D;
exports.LB_GETTOPINDEX          = 0x018E;
exports.LB_FINDSTRING           = 0x018F;
exports.LB_GETSELCOUNT          = 0x0190;
exports.LB_GETSELITEMS          = 0x0191;
exports.LB_SETTABSTOPS          = 0x0192;
exports.LB_GETHORIZONTALEXTENT  = 0x0193;
exports.LB_SETHORIZONTALEXTENT  = 0x0194;
exports.LB_SETCOLUMNWIDTH       = 0x0195;
exports.LB_ADDFILE              = 0x0196;
exports.LB_SETTOPINDEX          = 0x0197;
exports.LB_GETITEMRECT          = 0x0198;
exports.LB_GETITEMDATA          = 0x0199;
exports.LB_SETITEMDATA          = 0x019A;
exports.LB_SELITEMRANGE         = 0x019B;
exports.LB_SETANCHORINDEX       = 0x019C;
exports.LB_GETANCHORINDEX       = 0x019D;
exports.LB_SETCARETINDEX        = 0x019E;
exports.LB_GETCARETINDEX        = 0x019F;
exports.LB_SETITEMHEIGHT        = 0x01A0;
exports.LB_GETITEMHEIGHT        = 0x01A1;
exports.LB_FINDSTRINGEXACT      = 0x01A2;
exports.LB_SETLOCALE            = 0x01A5;
exports.LB_GETLOCALE            = 0x01A6;
exports.LB_SETCOUNT             = 0x01A7;
exports.LB_INITSTORAGE          = 0x01A8;
exports.LB_ITEMFROMPOINT        = 0x01A9;
exports.LB_MSGMAX               = 0x01B3; 
//exports.LB_MSGMAX             = 0x01B1; 
//exports.LB_MSGMAX             = 0x01B0; 
//exports.LB_MSGMAX             = 0x01A8] 

exports.CBN_ERRSPACE   = (-1);
exports.CBN_SELCHANGE  = 1;
exports.CBN_DBLCLK     = 2;
exports.CBN_SETFOCUS   = 3;
exports.CBN_KILLFOCUS  = 4;
exports.CBN_EDITCHANGE = 5;
exports.CBN_EDITUPDATE = 6;
exports.CBN_DROPDOWN   = 7;
exports.CBN_CLOSEUP    = 8;
exports.CBN_SELENDOK   = 9;
exports.CBN_SELENDCANCEL = 10;
exports.CB_GETEDITSEL            = 0x0140;
exports.CB_LIMITTEXT             = 0x0141;
exports.CB_SETEDITSEL            = 0x0142;
exports.CB_ADDSTRING             = 0x0143;
exports.CB_DELETESTRING          = 0x0144;
exports.CB_DIR                   = 0x0145;
exports.CB_GETCOUNT              = 0x0146;
exports.CB_GETCURSEL             = 0x0147;
exports.CB_GETLBTEXT             = 0x0148;
exports.CB_GETLBTEXTLEN          = 0x0149;
exports.CB_INSERTSTRING          = 0x014A;
exports.CB_RESETCONTENT          = 0x014B;
exports.CB_FINDSTRING            = 0x014C;
exports.CB_SELECTSTRING          = 0x014D;
exports.CB_SETCURSEL             = 0x014E;
exports.CB_SHOWDROPDOWN          = 0x014F;
exports.CB_GETITEMDATA           = 0x0150;
exports.CB_SETITEMDATA           = 0x0151;
exports.CB_GETDROPPEDCONTROLRECT = 0x0152;
exports.CB_SETITEMHEIGHT         = 0x0153;
exports.CB_GETITEMHEIGHT         = 0x0154;
exports.CB_SETEXTENDEDUI         = 0x0155;
exports.CB_GETEXTENDEDUI         = 0x0156;
exports.CB_GETDROPPEDSTATE       = 0x0157;
exports.CB_FINDSTRINGEXACT       = 0x0158;
exports.CB_SETLOCALE             = 345;
exports.CB_GETLOCALE             = 346;
exports.CB_GETTOPINDEX           = 347;
exports.CB_SETTOPINDEX           = 348;
exports.CB_GETHORIZONTALEXTENT   = 349;
exports.CB_SETHORIZONTALEXTENT   = 350;
exports.CB_GETDROPPEDWIDTH       = 351;
exports.CB_SETDROPPEDWIDTH       = 352;
exports.CB_INITSTORAGE           = 353;
exports.CB_MSGMAX                = 0x165; 
//exports.CB_MSGMAX                = 0x163; 
//exports.CB_MSGMAX                = 0x162; 
//exports.CB_MSGMAX                = 0x15B; 

exports.EN_SETFOCUS  = 0x0100;
exports.EN_KILLFOCUS = 0x0200;
exports.EN_CHANGE    = 0x0300;
exports.EN_UPDATE    = 0x0400;
exports.EN_ERRSPACE  = 0x0500;
exports.EN_MAXTEXT   = 0x0501;
exports.EN_HSCROLL   = 0x0601;
exports.EN_VSCROLL   = 0x0602;

exports.EM_GETSEL              = 0x00B0;
exports.EM_SETSEL              = 0x00B1;
exports.EM_GETRECT             = 0x00B2;
exports.EM_SETRECT             = 0x00B3;
exports.EM_SETRECTNP           = 0x00B4;
exports.EM_SCROLL              = 0x00B5;
exports.EM_LINESCROLL          = 0x00B6;
exports.EM_SCROLLCARET         = 0x00B7;
exports.EM_GETMODIFY           = 0x00B8;
exports.EM_SETMODIFY           = 0x00B9;
exports.EM_GETLINECOUNT        = 0x00BA;
exports.EM_LINEINDEX           = 0x00BB;
exports.EM_SETHANDLE           = 0x00BC;
exports.EM_GETHANDLE           = 0x00BD;
exports.EM_GETTHUMB            = 0x00BE;
exports.EM_LINELENGTH          = 0x00C1;
exports.EM_REPLACESEL          = 0x00C2;
exports.EM_GETLINE             = 0x00C4;
exports.EM_LIMITTEXT           = 0x00C5;
exports.EM_CANUNDO             = 0x00C6;
exports.EM_UNDO                = 0x00C7;
exports.EM_FMTLINES            = 0x00C8;
exports.EM_LINEFROMCHAR        = 0x00C9;
exports.EM_SETTABSTOPS         = 0x00CB;
exports.EM_SETPASSWORDCHAR     = 0x00CC;
exports.EM_EMPTYUNDOBUFFER     = 0x00CD;
exports.EM_GETFIRSTVISIBLELINE = 0x00CE;
exports.EM_SETREADONLY         = 0x00CF;
exports.EM_SETWORDBREAKPROC    = 0x00D0;
exports.EM_GETWORDBREAKPROC    = 0x00D1;
exports.EM_GETPASSWORDCHAR     = 0x00D2;
exports.EM_SETMARGINS          = 211;
exports.EM_GETMARGINS          = 212;
exports.EM_SETLIMITTEXT        = exports.EM_LIMITTEXT;    //win40 Name change
exports.EM_GETLIMITTEXT        = 213;
exports.EM_POSFROMCHAR         = 214;
exports.EM_CHARFROMPOS         = 215;
exports.EM_SETIMESTATUS        = 216;
exports.EM_GETIMESTATUS        = 217;

exports.SBM_SETPOS = 224;             
exports.SBM_GETPOS = 225;             
exports.SBM_SETRANGE = 226;           
exports.SBM_SETRANGEREDRAW = 230;     
exports.SBM_GETRANGE = 227;           
exports.SBM_ENABLE_ARROWS = 228;      
exports.SBM_SETSCROLLINFO = 233;
exports.SBM_GETSCROLLINFO = 234;
exports.SBM_GETSCROLLBARINFO = 235;   
exports.DM_GETDEFID = (exports.WM_USER+0);
exports.DM_SETDEFID = (exports.WM_USER+1);
exports.DM_REPOSITION = (exports.WM_USER+2);
exports.PSM_PAGEINFO = (exports.WM_USER+100);
exports.PSM_SHEETINFO = (exports.WM_USER+101);
